### ISSUE TITLE
Apparent race condition is causing failures

### ISSUE THREAD

**Issue by creswick** [jamesdbloom/mockserver#300]
Submitting requests quickly to a mock that processes requests through a callback class triggers NoSuchElement: null Exceptions stemming from RequestLogFilter.onRequest(...)'s used of the EvictingQueue.

Reproduction:

- Create a mock with a callback class that does some moderately time consuming task (such as reading a fixture from disk; or sleep for ~10 seconds before responding to test)
 - submit requests to that mock at a moderate frequency (15 second delays were sufficient in our app; using multiple threads to send concurrent requests, or using a load testing tool should be sufficient although I have not tried that.)
 - Observe exceptions of the form:

```
java.util.NoSuchElementException: null
	at java.util.ArrayDeque.removeFirst(ArrayDeque.java:280)
	at java.util.ArrayDeque.remove(ArrayDeque.java:447)
	at com.google.common.collect.EvictingQueue.add(EvictingQueue.java:105)
	at org.mockserver.filters.RequestLogFilter.onRequest(RequestLogFilter.java:41)
	at org.mockserver.filters.Filters.applyOnRequestFilters(Filters.java:33)
	at org.mockserver.mock.action.ActionHandler.processAction(ActionHandler.java:27)
	at org.mockserver.mockserver.MockServerHandler.channelRead0(MockServerHandler.java:179)
	at org.mockserver.mockserver.MockServerHandler.channelRead0(MockServerHandler.java:41)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:307)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:293)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHand
...
```

This appears to be happening because the filter and the queue's underlying that filter are not thread-safe, allowing multiple threads to pass the size check in EvictingQueue.add(E .){...} before invoking delegate.remove().  Thus, delegate.remove() is being called when delegate.size() == 0, triggering the exception seen above.

The workaround we've found is to slow the request frequency to (an essentially unusable) 1 request / minute.

I apologize for not submitting a minimal test example; ran out of time today while debugging this crash.  Since the RequestLogFilter has synchornized keywords (in comments) it seems like this is probably due to a recent change that may be easy to revert, or to finish the transition to making this class thread safe.

In short: I think you all may know how to fix this quickly an easily without a minimal example, if not, let me know and I'll try to cook one up.

**Comment by jamesdbloom** [comment]
Closing as now fixed, I don't believe the fix is in the current RELEASE version.  However I will issue a new RELEASE version the next few days once a TLS workout is complete to handle badly behaving TLS servers.

**Comment by creswick** [comment]
Thanks!

**Comment by rnavarropiris** [comment]
Does release 3.10.8 fix this?
And if not, in which version is the fix expected?

Thanks in advance! :)
